syntax = "proto3";
package pain.core.deva;

import "pain/core/common.proto";

option cc_generic_services = true;

service DevaService {
  rpc create(CreateRequest) returns (CreateResponse);
  rpc open(OpenRequest) returns (OpenResponse);
  rpc close(CloseRequest) returns (CloseResponse);
  rpc remove(RemoveRequest) returns (RemoveResponse);
  rpc seal(SealRequest) returns (SealResponse);

  rpc create_chunk(CreateChunkRequest) returns (CreateChunkResponse);
  rpc append_chunk(AppendChunkRequest) returns (AppendChunkResponse);
  rpc remove_chunk(RemoveChunkRequest) returns (RemoveChunkResponse);
  rpc seal_chunk(SealChunkRequest) returns (SealChunkResponse);
  rpc seal_and_new_chunk(SealAndNewChunkRequest)
      returns (SealAndNewChunkResponse);
}

message CreatePoolRequest {
  UUID uuid = 1;
}

message CreatePoolResponse {}

message CreateRequest {
  UUID pool = 1;
  string path = 2;
}

message CreateResponse {
  UUID uuid = 1;
}

message OpenRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message OpenResponse {
  FileMeta file_meta = 1;
}

message CloseRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message CloseResponse {}

message RemoveRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message RemoveResponse {}

message SealRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message SealResponse {}

message SealAndNewChunkRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message SealAndNewChunkResponse {
  UUID pool = 1;
  UUID uuid = 2;
}

message CreateChunkRequest {}

message CreateChunkResponse {
  UUID pool = 1;
  UUID uuid = 2;
}

message AppendChunkRequest {
  UUID pool = 1;
  UUID uuid = 2;
  uint64 offset = 3;
  uint32 length = 4;
  uint32 digest = 5;
}

message AppendChunkResponse {
  uint64 offset = 1;
}

message RemoveChunkRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message RemoveChunkResponse {}

message SealChunkRequest {
  UUID pool = 1;
  UUID uuid = 2;
}

message SealChunkResponse {}
